cosima_cookbook/database.py

```diff
@@ -578,86 +578,93 @@ def build_index(
 
         if force:
-            # Delete entries that we're reindexing
-            for f in files & existing_files:
-                existing_ncfile = (
-                    session.query(NCFile)
-                    .join(NCExperiment)
-                    .filter(NCExperiment.experiment == str(directory.name))
-                    .filter(NCExperiment.root_dir == str(directory.resolve()))
-                    .filter(NCFile.ncfile == f)
-                    .one_or_none()
-                )
-                if existing_ncfile is not None:
-                    print("Re-indexing {}".format(f))
-                    session.delete(existing_ncfile)
-            session.flush()
+            # Prune all files to force re-indexing data
+            _prune_files(expt, session, {}, delete=True)
         else:
-            # Only pass files that are not already in DB
-            files = files - existing_files
+            # Prune files that exist in the database but are not present on disk
+            _prune_files(expt, session, files, delete=(prune == "delete"))
 
-        indexed += index_experiment(files, session, expt, client)
+        if len(files) > 0:
+            if len(expt.ncfiles) > 0:
+                # Only pass files that are not already in DB
+                files = {
+                    f.ncfile
+                    for f in session.query(NCFile)
+                    .with_parent(expt)
+                    .filter(NCFile.ncfile.notin_(files))
+                }
 
-        # if everything went smoothly, commit these changes to the database
-        session.commit()
+            indexed += index_experiment(files, session, expt, client)
+
+            # if everything went smoothly, commit these changes to the database
+            session.commit()
 
     return indexed
 
 
-def prune_files(expt, session, files, delete=True):
-    """Delete or mark as not present the database entries for specified
-    files objects within the given experiment
+def _prune_files(expt, session, files, delete=True):
+    """Delete or mark as not present the database entries that are
+    not present in the list of files
     """
-    for f in files:
+    if len(expt.ncfiles) == 0:
+        # No entries in DB, special case just return as there is nothing
+        # to prune and trying to do so will raise errors
+        return
+
+    # Missing are physically missing from disk, or where marked as not
+    # present previously. Can also be a broken file which didn't index
+    missing_ncfiles = (
+        session.query(NCFile)
+        .with_parent(expt)
+        .filter(NCFile.ncfile.notin_(files) | (NCFile.present == False))
+    )
+    if missing_ncfiles.first() is not None:
         if delete:
-            session.delete(f)
+            missing_ncfiles.delete(synchronize_session=False)
         else:
-            f.present = False
+            missing_ncfiles.update({NCFile.present: False}, synchronize_session=False)
+
+        session.commit()
```

Comment:
I haven't tested this, so feel free to ignore. I don't know what's breaking without the `.commit()`, but it's probably due to `synchronize_session`. It might be possible to use `session.expire_all()` between lines 633 and 634, to expire existing persistent objects (things like `expt`), so that they're re-queried on next use.

Reply:
The specific test that fails is this:
```python
    def test_reindex_force(session_db):
        session, db = session_db
        database.build_index("test/data/indexing/broken_file", session)
        assert db.check()
    
        # re-run the index, make sure re-index
        reindexed = database.build_index(
            "test/data/indexing/broken_file", session, force=True
        )
>       assert reindexed == 1
E       assert 0 == 1
```

Your suggestion for `session.expire_all()` works, so I'll go with that.

Reply:
I do get this warning though regardless if I use `commit` or `expire_all`
```python
test/test_indexing.py::test_reindex_force
  /home/502/aph502/.local/lib/python3.9/site-packages/sqlalchemy/orm/session.py:1952: SAWarning: Identity map already had an identity for (<class 'cosima_cookbook.database.NCFile'>, (1,), None), replacing it with newly flushed object.   Are there load operations occurring inside of an event handler within the flush?
```